File: control-plane/tests/edition.test.ts
import test from 'node:test';
import assert from 'node:assert/strict';
import crypto from 'node:crypto';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { ApiError } from '../lib/api.ts';
import { activeEdition, hasFeature, requireFeature, resetEditionCacheForTests, validateLicense } from '../lib/edition.ts';

const EDITION_ENV = '2PAPI_EDITION';
const PUBKEY_ENV = '2PAPI_LICENSE_PUBKEY';

function makeKeypair() {
  const { publicKey, privateKey } = crypto.generateKeyPairSync('ed25519');
  const spki = publicKey.export({ type: 'spki', format: 'der' }) as Buffer;
  return { pubHex: spki.subarray(spki.length - 32).toString('hex'), privateKey };
}

function signLicense(privateKey: crypto.KeyObject, prefix: string, payload: Record<string, unknown>) {
  const b64payload = Buffer.from(JSON.stringify(payload), 'utf8').toString('base64url');
  const msg = Buffer.from(`${prefix}.${b64payload}`, 'utf8');
  const sig = crypto.sign(null, msg, privateKey);
  return `${prefix}:${b64payload}.${sig.toString('base64url')}`;
}

// Minimal structural type for the test context we use (t.after cleanup).
interface Cleanup {
  after(fn: () => void | Promise<void>): void;
}

function withEnv(t: Cleanup, key: string, value: string | undefined) {
  const old = process.env[key];
  t.after(() => { if (old === undefined) delete process.env[key]; else process.env[key] = old; });
  if (value === undefined) delete process.env[key]; else process.env[key] = value;
}

// Each license-file test gets a fresh cwd + clean gate cache.
function withLicenseDir(t: Cleanup) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), '2papi-edition-'));
  const originalCwd = process.cwd();
  process.chdir(dir);
  resetEditionCacheForTests();
  t.after(() => {
    process.chdir(originalCwd);
    fs.rmSync(dir, { recursive: true, force: true });
    resetEditionCacheForTests();
  });
  return dir;
}

test('no env and no license degrades to OSS; gated paths refuse', (t) => {
  withEnv(t, EDITION_ENV, undefined);
  assert.equal(activeEdition(), 'oss');
  assert.equal(hasFeature('orgs'), false);
  assert.equal(hasFeature('audit_export'), false);
  assert.throws(() => requireFeature('orgs'), (e: unknown) => e instanceof ApiError && e.status === 403 && e.code === 'feature_not_licensed');
});

test('env override wins: ent unlocks all known features, garbage degrades', async (t) => {
  withEnv(t, EDITION_ENV, undefined);
  assert.equal(activeEdition(), 'oss');
  assert.equal(hasFeature('orgs'), false);

  withEnv(t, EDITION_ENV, 'ent');
  assert.equal(activeEdition(), 'ent');
  assert.equal(hasFeature('orgs'), true);
  assert.equal(hasFeature('sso'), true);

  withEnv(t, EDITION_ENV, 'cloud');
  assert.equal(activeEdition(), 'cloud');

  withEnv(t, EDITION_ENV, 'garbage');
  assert.equal(activeEdition(), 'oss');

  // Unknown features never unlock even under an explicit override.
  withEnv(t, EDITION_ENV, 'ent');
  assert.equal(hasFeature('nonexistent'), false);
});

test('signed ent license unlocks only its feature list', async (t) => {
  withEnv(t, EDITION_ENV, undefined);
  const { pubHex, privateKey } = makeKeypair();
  withEnv(t, PUBKEY_ENV, pubHex);
  withLicenseDir(t);
  const lic = signLicense(privateKey, 'ent', { ed: 'ent', cid: 'acme', cap: 50000, iat: 1, exp: Math.floor(Date.now() / 1000) + 3600, f: ['orgs'] });
  fs.writeFileSync('2papi.license', lic + '\n');
  resetEditionCacheForTests();

  assert.equal(activeEdition(), 'ent');
  assert.equal(hasFeature('orgs'), true);
  assert.equal(hasFeature('sso'), false, 'features not in the license stay locked');

  assert.throws(() => requireFeature('sso'), (e: unknown) => e instanceof ApiError && e.status === 403 && e.code === 'feature_not_licensed');
});

test('expired / not-yet / swapped-prefix / wrong-key licenses degrade to OSS', async (t) => {
  withEnv(t, EDITION_ENV, undefined);
  const { pubHex, privateKey } = makeKeypair();
  withEnv(t, PUBKEY_ENV, pubHex);
  withLicenseDir(t);
  const nowSec = Math.floor(Date.now() / 1000);

  fs.writeFileSync('2papi.license', signLicense(privateKey, 'ent', { ed: 'ent', exp: nowSec - 10 }));
  resetEditionCacheForTests();
  assert.equal(activeEdition(), 'oss', 'expired');

  fs.writeFileSync('2papi.license', signLicense(privateKey, 'ent', { ed: 'ent', exp: nowSec + 3600, nbf: nowSec + 60 }));
  resetEditionCacheForTests();
  assert.equal(activeEdition(), 'oss', 'not valid yet');

  // Prefix says ent, payload signed as cloud — signature check must fail.
  fs.writeFileSync('2papi.license', signLicense(privateKey, 'cloud', { ed: 'cloud', exp: nowSec + 3600 }).replace(/^cloud:/, 'ent:'));
  resetEditionCacheForTests();
  assert.equal(activeEdition(), 'oss', 'swapped prefix breaks the signature');

  const other = makeKeypair();
  fs.writeFileSync('2papi.license', signLicense(other.privateKey, 'ent', { ed: 'ent', exp: nowSec + 3600 }));
  resetEditionCacheForTests();
  assert.equal(activeEdition(), 'oss', 'wrong signing key');
});

test('validateLicense rejects malformed input explicitly', () => {
  assert.throws(() => validateLicense('garbage', 'a'.repeat(64)), /format|key/i);
  assert.throws(() => validateLicense('ent:abc.def', ''), /no trusted public key/i);
  assert.throws(() => validateLicense(`ent:${Buffer.from('{}').toString('base64url')}.aa`, 'zz'), /key/i);
});

File: internal/edition/edition_test.go
package edition

import (
	"crypto/ed25519"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
	"time"

	"github.com/Rethinger/2papi/internal/license"
)

func withEnv(t *testing.T, val string) {
	t.Helper()
	t.Setenv(EnvVar, val)
}

func TestActiveDefaultsToOSS(t *testing.T) {
	withEnv(t, "")
	dir := t.TempDir()
	chdir(t, dir)
	if got := Active(); got != OSS {
		t.Fatalf("no env, no license: want %q got %q", OSS, got)
	}
}

func TestActiveEnvWinsAndUnknownFallsBackToOSS(t *testing.T) {
	withEnv(t, "cloud")
	if !IsCloud() {
		t.Fatal("env cloud should activate cloud")
	}
	withEnv(t, "ENT")
	if !IsEnterprise() {
		t.Fatal("env ENT should activate enterprise")
	}
	withEnv(t, "garbage")
	if !IsOSS() {
		t.Fatal("unknown env must degrade to OSS")
	}
}

// signed writes a properly Ed25519-signed license file and returns it.
func signed(t *testing.T, priv ed25519.PrivateKey, ed string, days int) string {
	t.Helper()
	now := time.Now()
	payload := map[string]any{
		"ed": ed, "cid": "acme", "cap": 1000,
		"iat": now.Unix(), "exp": now.Unix() + int64(days)*86400,
		"f": []string{"sso", "orgs"},
	}
	raw, err := json.Marshal(payload)
	if err != nil {
		t.Fatal(err)
	}
	b64p := base64.RawURLEncoding.EncodeToString(raw)
	sig := ed25519.Sign(priv, []byte(ed+"."+b64p))
	return ed + ":" + b64p + "." + base64.RawURLEncoding.EncodeToString(sig)
}

func TestActiveSignedLicenseFile(t *testing.T) {
	withEnv(t, "") // license file path decides
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatal(err)
	}
	t.Setenv(license.PubKeyEnv, hexEncode(pub))

	dir := t.TempDir()
	chdir(t, dir)

	write := func(content string) {
		if err := os.WriteFile(LicenseFile, []byte(content), 0o600); err != nil {
			t.Fatal(err)
		}
	}

	write(signed(t, priv, "cloud", 30))
	if got := Active(); got != Cloud {
		t.Fatalf("signed cloud license: want %q got %q", Cloud, got)
	}

	write(signed(t, priv, "ent", 365))
	if got := Active(); got != ENT {
		t.Fatalf("signed ent license: want %q got %q", ENT, got)
	}

	// Expired license must degrade to OSS.
	expired := signed(t, priv, "cloud", -1)
	if err := os.WriteFile(LicenseFile, []byte(expired), 0o600); err != nil {
		t.Fatal(err)
	}
	if got := Active(); got != OSS {
		t.Fatalf("expired license: want %q got %q", OSS, got)
	}

	// Garbage file must NOT unlock anything.
	write("free tokens!!")
	if got := Active(); got != OSS {
		t.Fatalf("garbage license: want %q got %q", OSS, got)
	}

	// A license signed by a DIFFERENT key must be rejected.
	_, otherPriv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatal(err)
	}
	write(signed(t, otherPriv, "cloud", 30))
	if got := Active(); got != OSS {
		t.Fatalf("foreign-key license: want %q got %q", OSS, got)
	}

	// No pubkey configured at all: even a valid signature is untrusted.
	os.Unsetenv(license.PubKeyEnv)
	defer os.Setenv(license.PubKeyEnv, hexEncode(pub))
	write(signed(t, priv, "cloud", 30))
	if got := Active(); got != OSS {
		t.Fatalf("no trusted key: want %q got %q", OSS, got)
	}
}

func hexEncode(k ed25519.PublicKey) string {
	return hex.EncodeToString(k)
}

func chdir(t *testing.T, dir string) {
	t.Helper()
	old, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	if err := os.Chdir(dir); err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = os.Chdir(old) })
}
